### DISCUSSION TITLE
Getting unintended panic

### DISCUSSION THREAD

**Discussion by Vaimer9** [zesterer/chumsky#128] [Q&A] [answered]
I've found time to get working again and found myself stuck another problem. After [changing](https://gist.github.com/Vaimer9/d354e50accb72c89da5dab1cfd367944/revisions) the [source code](https://gist.github.com/Vaimer9/d354e50accb72c89da5dab1cfd367944) to work for multiple lines, I was getting an unintended panic:
![image](https://user-images.githubusercontent.com/81609819/164007241-b32dd25a-7e6b-4251-8d7c-df3862306915.png)
for the following input:
```
fn firstfunction(a, b, c) {
    echo line_one
    echo line_two
}

fn secondfunction(b) {
    echo extras
}
```
I was wondering if you could let me know why this was happening

**Comment by zesterer** [comment, marked as answer]
This panic happens because the parser matched a repeating pattern without consuming any input. This is almost always a bug in the parser (and it's common enough that I chose to add a panic for it). For example:

```rs
let a_chain = just('a').repeated();

let a_chain_chain = a_chain.repeated();
```

This parser will loop infinitely on the input `b` because the first parser, `a_chain`, is permitted to consume none of the input, and `a_chain_chain` looks for that input many times, leading to an infinite loop.

In your parser, you have  `.or_not().repeated()`. This is highly suspect: `or_not` is permitted to consume no inputs (as in the example I gave) and `repeated` will look for that pattern many times. I would suggest refactoring this.

**Comment by Vaimer9** [comment]
Makes sense! removing the `.or_not()` behind the repeated removes the panic, although the output I get is not expected:
```
Ok(
    Fn {
        name: "firstfunction",
        args: [
            "a",
            "b",
            "c",
        ],
        body: Some(
            [
                Command(
                    "echo line_one echo line_two",
                ),
            ],
        ),
        then: Some(
            Fn {
                name: "secondfunction",
                args: [
                    "b",
                ],
                body: Some(
                    [
                        Command(
                            "echo extras",
                        ),
                    ],
                ),
                then: None,
            },
        ),
    },
)
```
I needed there to be 2 `Command`s rather than a single `Command` comprising of two lines.
I thought `take_until(just('\n')` would be helpful but I can't do that as I need `expr` rather than just a string, my next move was trying out `then_ignore(just('\n'))` but that also raised unexpected behavior. What should I do?

**Reply by zesterer** [reply]
The `.padded()` combinator will consume newlines too. If this is not desirable, I'd recommend something like `.padded_by(one_of(" \t"))` instead.

**Reply by Vaimer9** [reply]
which `.padded()` is in question here?

**Reply by zesterer** [reply]
`text::ident().padded().repeated()` in `cmd`. This will keep reading identifiers, even after a newline.

**Reply by Vaimer9** [reply]
ok, so I made [some changes](https://github.com/Vaimer9/vsh/blob/rework/crates/parser/src/lib.rs#L24) (its all in a repo now) the output is just:
```
 Err(
    [
        Simple {
            span: 32..33,
            reason: Unexpected,
            expected: {
                Some(
                    ';',
                ),
                Some(
                    '$',
                ),
                Some(
                    '"',
                ),
                Some(
                    '0',
                ),
            },
            found: Some(
                'e',
            ),
            label: None,
        },
    ],
)
```
you can find/track the input [here](https://github.com/Vaimer9/vsh/blob/rework/main.vsh)

**Reply by Vaimer9** [reply]
I wanted semicolons rather than tabs

**Reply by zesterer** [reply]
Just so you're aware, `.padded_by(whitespace())` is the same thing as `.padded()`. The `whitespace` parser also recognises newlines.

**Reply by zesterer** [reply]
Additionally, `padded_by` recognises `<padding><inner><padding>`. You probably want `.then_ignore(just(';'))` to recognise a trailing semicolon.

**Reply by Vaimer9** [reply]
That does not seem to help either, although I get your point with adding the `.then_ignore()`

**Reply by Vaimer9** [reply]
nevermind it runs perfectly now, misplaced the `delimited_by()`
